bookkeeper-server/src/main/java/org/apache/bookkeeper/http/ListLedgerService.java

```diff
@@ -0,0 +1,177 @@
+      implements BookkeeperInternalCallbacks.GenericCallback<LedgerMetadata> {
+        final long ledgerId;
+
+        ReadLedgerMetadataCallback(long ledgerId) {
+            this.ledgerId = ledgerId;
+        }
+
+        long getLedgerId() {
+            return ledgerId;
+        }
+
+        public void operationComplete(int rc, LedgerMetadata result) {
+            if (rc != 0) {
+                setException(BKException.create(rc));
+            } else {
+                set(result);
+            }
+        }
+    }
+    static void keepLedgerMetadata(ReadLedgerMetadataCallback cb, LinkedHashMap<String, String> output) throws Exception {
+        LedgerMetadata md = cb.get();
+        output.put(Long.valueOf(cb.getLedgerId()).toString(), new String(md.serialize(), UTF_8));
+    }
+
+    @Override
+    public HttpServiceResponse handle(HttpServiceRequest request) throws Exception {
+        HttpServiceResponse response = new HttpServiceResponse();
+        // GET
+        // parameter could be like: print_metadata=true&page=PageIndex
+        if (HttpServer.Method.GET == request.getMethod()) {
+            Map<String, String> params = request.getParams();
+            // default not print metadata
+            boolean printMeta = (params != null) &&
+              params.containsKey("print_metadata") &&
+              params.get("print_metadata").equals("true");
+
+            // Page index should start from 1;
+            int pageIndex = (printMeta && params.containsKey("page")) ?
+              Integer.parseInt(params.get("page")) :
+              -1;
+
+            ZooKeeper zk = ZooKeeperClient.newBuilder()
+              .connectString(conf.getZkServers())
+              .sessionTimeoutMs(conf.getZkTimeout())
+              .build();
+            LedgerManagerFactory mFactory = LedgerManagerFactory.newLedgerManagerFactory(conf, zk);
+            LedgerManager manager = mFactory.newLedgerManager();
+            LedgerManager.LedgerRangeIterator iter = manager.getLedgerRanges();
+
+            // output <ledgerId: ledgerMetadata>
+            LinkedHashMap<String, String> output = Maps.newLinkedHashMap();
+            // futures for readLedgerMetadata for each page.
+            List<ReadLedgerMetadataCallback> futures = Lists.newArrayListWithExpectedSize(LIST_LEDGER_BATCH_SIZE);
+
+            if (printMeta) {
+                int ledgerIndex = 0;
+
+                // start and end ledger index for wanted page.
+                int startLedgerIndex = 0;
+                int endLedgerIndex = 0;
+                if(pageIndex > 0) {
+                    startLedgerIndex = (pageIndex - 1) * LIST_LEDGER_BATCH_SIZE;
+                    endLedgerIndex = startLedgerIndex + LIST_LEDGER_BATCH_SIZE - 1;
+                }
+
+                // get metadata
+                while (iter.hasNext()) {
+                    LedgerManager.LedgerRange r = iter.next();
```

Comment:
That is a good idea. But seems we only provide a getLedgerRanges api to read all ledgerids out. and it will only call zk once to get it all. 
The paging here is mainly to avoid readLedgerMetadata calls, since it is one ledger one zk call. the paging has avoid this a lot. 
If user want to read few number of ledgers, and they know the ledgerid, they could use getLedgerMetadata service.

Reply:
I think paging might be need fair amount of work to implement it, which might also require changes to ledger manager interface. I would suggest:

- removing list ledgers from current work
- create a separate issue for implementing list ledgers in a paging way.

how does that sound?

Reply:
^ this can be collaborated with @eolivelli 's idea on metadata filtering.

Reply:
Thanks, opened issue #542 for tracking it. and would like to keep current changes, then later we will only need add needed change.

Reply:
this is fine with me